File: algorithm-design/hackerrank/bfsshortreach/bfsshortreach.py
#!/usr/bin/env python3
# https://www.hackerrank.com/challenges/bfsshortreach

import collections
import os
import unittest

from typing import List


# pylint: disable=W0613
def bfs(n: int, m: int, edges: List[List[int]], s: int) -> List[int]:
    adjacency = [[False] * n for _ in range(n)]
    for edge in edges:
        adjacency[edge[0] - 1][edge[1] - 1] = True
        adjacency[edge[1] - 1][edge[0] - 1] = True
    queue = collections.deque()
    queue.append((s - 1, 0))
    distances = [-1] * n
    while queue:
        vertex, distance = queue.pop()
        if distances[vertex] == -1:
            distances[vertex] = distance
            for i in range(n):
                if adjacency[vertex][i]:
                    queue.append((i, distance + 6))
    del distances[s - 1]
    return distances


class TestCode(unittest.TestCase):
    # pylint: disable=R0914
    def runner(self, name):
        io_lines = [[[]]] * 2
        for index, template in enumerate(['input%s.txt', 'output%s.txt']):
            path = os.path.join(os.path.split(__file__)[0], template % name)
            with open(path, 'r') as handle:
                lines = handle.readlines()
            io_lines[index] = [line.strip().split(' ') for line in lines]
        tests = int(io_lines[0][0][0])
        offset = 1
        for test in range(tests):
            n = int(io_lines[0][offset][0])
            m = int(io_lines[0][offset][1])
            edges = io_lines[0][offset + 1:offset + 1 + m]
            edges = [[int(item) for item in row] for row in edges]
            s = int(io_lines[0][offset + 1 + m][0])
            offset += 1 + m + 1
            result = bfs(n, m, edges, s)
            expected = [int(item) for item in io_lines[1][test]]
            self.assertEqual(expected, result)

    def test_example(self):
        self.runner('_example')

File: algorithms/hackerrank/bfsshortreach/bfsshortreach.go
package bfsshortreach

// https://www.hackerrank.com/challenges/bfsshortreach

type entry struct {
	vertex   int32
	distance int32
}

// Bfs - implements the solution to the problem
func Bfs(n int32, m int32, edges [][]int32, s int32) []int32 {
	adjacency := make([][]bool, n)
	for i := 0; i < int(n); i++ {
		adjacency[i] = make([]bool, n)
		for j := 0; j < int(n); j++ {
			adjacency[i][j] = false
		}
	}
	for _, edge := range edges {
		adjacency[edge[0]-1][edge[1]-1] = true
		adjacency[edge[1]-1][edge[0]-1] = true
	}
	queue := make([]entry, 0)
	queue = append(queue, entry{s - 1, 0})
	distances := make([]int32, n)
	for i := 0; i < int(n); i++ {
		distances[i] = -1
	}
	for len(queue) > 0 {
		item := queue[0]
		queue = queue[1:]
		if distances[item.vertex] == -1 {
			distances[item.vertex] = item.distance
			for i := 0; i < int(n); i++ {
				if adjacency[int(item.vertex)][i] {
					queue = append(queue, entry{int32(i), item.distance + 6})
				}
			}
		}
	}
	result := make([]int32, n-1)
	index := 0
	for i := 0; i < int(n); i++ {
		if i != int(s-1) {
			result[index] = distances[i]
			index++
		}
	}
	return result
}
